### ISSUE TITLE
Time-scaled SLiM simulations have systematically fewer num_nodes and num_edges

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/stdpopsim#444]
See #442.

This is probably expected for time-scaled discrete-generation simulations, but we should satisfy ourselves that there isn't another underlying problem here. And is this problematic in general?

**Comment by petrelharp** [comment]
Are these simplified?

**Comment by petrelharp** [comment]
By 'time-scaled', you mean Q != 1?

**Comment by grahamgower** [comment]
Simplified, yes: https://github.com/popsim-consortium/stdpopsim/blob/master/stdpopsim/slim_engine.py#L533-L541

And yes, I mean Q>1 (Q=10 in the cases I looked at).

**Comment by petrelharp** [comment]
Well, fewer nodes would imply fewer edges, since each node has to be connected to the other ones. And I think fewer nodes makes sense - with fewer individuals, there will be more re-use of nodes, even with the same amount of coalescence. Imagine taking the set of marginal trees; with no rescaling and a very big population imagine that every node in every marginal tree is distinct; but as the total population size goes down (and since time changes also, the number of individuals-alive-ever goes down with Q^2) those trees get squooshed onto fewer distinct ancestral genomes.

Come to think of it, this gives us fodder for the "rescaling is not actually equivalent" bit in the manuscript, so maybe `num_nodes` *should* be a plot in the comparison?

**Comment by jeromekelleher** [comment]
> Come to think of it, this gives us fodder for the "rescaling is not actually equivalent" bit in the manuscript, so maybe num_nodes should be a plot in the comparison?

Makes sense to me - the trees are what matter at the end of the day.

**Comment by petrelharp** [comment]
> Makes sense to me - the trees are what matter at the end of the day.

And this is kinda tricky - because presumably the *trees* are good (marginally), but there's more correlation between trees along the sequence than you'd expect, somehow.

**Comment by petrelharp** [comment]
More re-using of nodes between non-adjacent trees was what I was thinking. Since density of recombinations on the trees is scaled to be the same, mean edge length and number of trees should be (almost?) the same.  But this is hard to think about.

**Comment by jeromekelleher** [comment]
That's interesting, I'm curious to take a look now. Can you point me to a set of plots from a SLiM simulation with Q=1 and Q>1 for comparison please @grahamgower? I've gotten a bit lost with all the ones that are going around.

**Comment by grahamgower** [comment]
In the latest validation plots (in #418), all the `slim1` simulations are with the new defaults. Ie. Q=1, burn_in=10. `onepop_*_slim2` is with burn_in=0, `onepop_*_slim3` is with Q=10. `twopop_*slim2` is with burn_in=0 and Q=10. Sorry, the slim2/slim3 numbers are a little inconsistent between onepop and twopop models, so maybe just look at the onepop models.

**Comment by grahamgower** [comment]
What are the implications here for tree sequence inference from empirical data? Presumably num_nodes and num_edges are non-identifiable, given genotype data?

**Comment by petrelharp** [comment]
Huh, my explanation of why more nodes => more edges doesn't agree with why I said there should be more nodes. =)

> num_nodes and num_edges are non-identifiable, given genotype data?

I think there's no obvious way to infer them, but different numbers does imply some differences in correlations along the genome that should in principle be detectable (although depending on what's going on, might be quite hard?))

**Comment by grahamgower** [comment]
I wonder if a neutral network can identify the difference between Q=1 and Q=10 from genotypes alone...

**Comment by andrewkern** [comment]
> Well, fewer nodes would imply fewer edges, since each node has to be connected to the other ones. And I think fewer nodes makes sense - with fewer individuals, there will be more re-use of nodes, even with the same amount of coalescence. Imagine taking the set of marginal trees; with no rescaling and a very big population imagine that every node in every marginal tree is distinct; but as the total population size goes down (and since time changes also, the number of individuals-alive-ever goes down with Q^2) those trees get squooshed onto fewer distinct ancestral genomes.

yes this is really interesting to think about- the size of the containing genealogy is shrinking, even if the parameters are scaled up. maybe this would show up in patterns of LD across genomes? would we perhaps expect longer range correlations to persist?

**Comment by dschride** [comment]
Oops, clicked the wrong button!

But what I meant to do was ask if whether this "sharing of nodes" problem is really just another way of saying that N is no longer >> n?

**Comment by grahamgower** [comment]
It's certainly an artifact of discretisation. I reckon there's an analogy here with sampling theory in signal processing (https://en.wikipedia.org/wiki/Nyquist%E2%80%93Shannon_sampling_theorem), if we think of the discrete generations as a sampling of continuous-time events. So our continuous-time "signal" has frequency at time `t` corresponding to the frequency of (coalescent, recombination) events at time `t`. I think this is a useful analogy, because we should be able to calculate the *Nyquist frequency* and thus determine when the discretisation is too coarse.

**Comment by grahamgower** [comment]
This also predicts that we should be able to do more aggressive scaling in the deep past, where events are less frequent.

**Comment by jeromekelleher** [comment]
Here's the ts properties for onepop_bottleneck_msprime1__onepop_bottleneck_slim3 (Q=10):

![Screenshot from 2020-04-01 10-10-11](https://user-images.githubusercontent.com/2664569/78120002-2d2dff80-7401-11ea-8461-37360d45c8f9.png)

**Comment by jeromekelleher** [comment]
I'm just going to state the obvious here:

- num_trees and num_sites are have the same distribution, and therefore the total branch lengths of the trees are the same (since recombinations and mutations fall proportional to branch lengths)
- We have systematically fewer nodes and edges in the SLiM sims. Probably the simplest explanation for this is that the SLiM sims have lots of non-binary nodes. This makes sense, right - smaller N means larger possibility of multiple mergers? 

We can test this easily enough by looking at the arity distribution:
```python
node_arity = []
for tree in ts.trees():
    for node in tree.nodes():
           if tree.isinternal(node):
                 node_arity.append(tree.num_children(node))
```

The msprime, the distribution will be very boring, but for SLiM we should see it being skewed away from 2 as we make Q larger.

**Comment by grahamgower** [comment]
Q=1 (slim1) versus Q=10 (slim3).
![node_arity](https://user-images.githubusercontent.com/8665077/78135712-a2f69300-7422-11ea-8adf-dd1a8b1fc9af.png)

**Comment by jeromekelleher** [comment]
Nice, thanks @grahamgower! So, trees get increasingly more starlike for Q > 1. That's a pretty simple and clear conclusion, right?

**Comment by grahamgower** [comment]
Yeah. But they're not going to get much more starlike unless the (scaled) population size is small. So, for drosophila with Q=1000, nodes with more than 2 children are still going to be very uncommon.

**Comment by jeromekelleher** [comment]
Right. I'm sure there's some coalescent theory we can draw on there, this is basically the n << N assumption as @dschride said above.

**Comment by grahamgower** [comment]
Well I'm quite satisfied that this is not something to be "fixed" in the slim engine. Unless anyone has anything to add, I think this issue can be closed.

**Comment by jeromekelleher** [comment]
Agreed. The only question really is how we present this in the manuscript, but perhaps we should move the discussion over there then.

**Comment by andrewkern** [comment]
sure let's move it there. we should also be sure to remember to note this in the documentation somewhere

**Comment by dschride** [comment]
> 
> 
> This also predicts that we should be able to do more aggressive scaling in the deep past, where (observable) events are less frequent.

You may be on to something here. Anyone know if something like this has been done before? But yeah, let's close and discuss elsewhere.

**Comment by petrelharp** [comment]
> This also predicts that we should be able to do more aggressive scaling in the deep past, where (observable) events are less frequent.

This is interesting! But I'm not sure - I think the relevant thing is the number of distinct ancestors, and since ancestral blocks break up into smaller chunks as you go back, this number can actually increase for a while before it decreases, I think?

**Comment by grahamgower** [comment]
> I wonder if a neutral network can identify the difference between Q=1 and Q=10 from genotypes alone...

To answer my own thought bubble, I did 5000 simulations of both onepop_bottleneck_slim1 (Q=1) and onepop_bottleneck_slim3 (Q=10) and tried to distinguish genotype matrixes from the two sets using an exchangable CNN architecture similar to that of Chan et al. (https://arxiv.org/abs/1802.06153). I only tried a few different hyperparameters, but the network was unable to learn, always stuck at 50% accuracy. This is a reasonably general-purpose architecture that readily learns other population parameters, so not learning probably indicates that the problem is very difficult or requires a larger genomic interval than 250 kb .

CNN code [here](https://github.com/grahamgower/tsnn) if anyone wants to play. Sorry, it's still pretty raw and undocumented, as I'm only playing with this in my spare time.
